### PR TITLE
Fix IP address collection with IPv6 disabled

### PR DESCRIPTION
**Pull request by m-horky**
When IPv6 is disabled, this code crashed Python with segfault, instead of rising an exception. This patch iterates over interfaces found by generic 'getaddrinfo()' call instead of querying v4 and v6 individually.

### CODE REVIEW

File: src/rhsmlib/facts/hwprobe.py [hunk 1]
```diff
@@ -782,20 +783,29 @@ def get_network_info(self) -> Dict[str, str]:
...
-            try:
-                info: List[tuple] = socket.getaddrinfo(hostname, None, socket.AF_INET6, socket.SOCK_STREAM)
-                ip_list = set([x[4][0] for x in info])
-                net_info["network.ipv6_address"] = ", ".join(ip_list)
-            except Exception as err:
-                log.debug("Error during resolving IPv6 address of hostname: %s, %s" % (hostname, err))
+            if ipv6_addresses:
+                net_info["network.ipv6_address"] = ", ".join(ipv6_addresses)
+            else:
+                log.debug(f"Error getting IPv6 addresses of {hostname} using socket inspection.")
```
Comment (jirihnidek) [inline src/rhsmlib/facts/hwprobe.py:808]: I would not scare users with any message containing word: "Error"

File: src/rhsmlib/facts/hwprobe.py [hunk 2]
```diff
@@ -782,20 +783,29 @@ def get_network_info(self) -> Dict[str, str]:
...
                 net_info["network.ipv6_address"] = ", ".join(self._get_ipv6_addr_list())
```
Comment (jirihnidek) [inline src/rhsmlib/facts/hwprobe.py:809]: When IPv6 is completely disabled, then the list of addresses still contains `::1` address, which is not correct. When IPv6 is disabled, then there should not be any IPv6 address in the list.

**Review by jirihnidek — CHANGES_REQUESTED** [review]
Thanks for the PR. I miss some unit tests for this case and instructions for disabling/enabling IPv6 in the commit message or comment of PR. Something like:

```
# Disable IPv6
sudo sysctl -w net.ipv6.conf.all.disable_ipv6=1
sudo sysctl -w net.ipv6.conf.default.disable_ipv6=1
```


```
# Enable IPv6 again
sudo sysctl -w net.ipv6.conf.all.disable_ipv6=0
sudo sysctl -w net.ipv6.conf.default.disable_ipv6=0
```

And I have some other requests. More details in comments.

File: src/rhsmlib/facts/hwprobe.py (hunk 2 above)
Comment (jirihnidek) [reply src/rhsmlib/facts/hwprobe.py:809]: It is strange. I can see: `network.ipv6_address: ::1` in system facts, because of this:

https://github.com/candlepin/subscription-manager/blob/main/src/rhsmlib/facts/hwprobe.py#L747

It is strange, because loopback interface contains only IPv4 addresses:

```
net.interface.lo.ipv4_address: 127.0.0.1
net.interface.lo.ipv4_address_list: 127.0.0.1
net.interface.lo.ipv4_broadcast: Unknown
net.interface.lo.ipv4_broadcast_list: Unknown
net.interface.lo.ipv4_netmask: 8
net.interface.lo.ipv4_netmask_list: 8
```

It was probably me, who has implemented that, but I think that it is wrong now, because we should not fabricate any non-existent facts in the code.

Comment (jirihnidek) [reply src/rhsmlib/facts/hwprobe.py:809]: Maybe we could fix it in other PR. It is only little bit related to this case. When I was talking about unit testing, then I had this `network.ipv6_address: ::1` in my mind.

**Comment by ptoscano** [comment]
TL;DR: I don't think we should include this patch.

I did some testing of this patch, and I noticed the log message `Could not obtain IPv6 addresses using socket inspection. Using ethtool instead.`, whereas its "current equivalent" (`Error during resolving IPv6 address of hostname: ...`) is not shown. Digging a bit more, I found why, and it's due to how `getaddrinfo()` works. The initial `getaddrinfo()` (the one used to find out the FQDN) is called with `socket.AF_UNSPEC` as the family type; according to the documentation [1]:

> The value AF_UNSPEC indicates that getaddrinfo() should return socket addresses for any address family (either IPv4 or IPv6, for example) that can be used with node and service.

The "either ... or" part is the key here: on a system that has both IPv4 and IPv6 set up, IPv4 is found as usable, and thus all the results of `getaddrinfo()` are _only_ `socket.AF_INET` results. This means that, in practically all the setups, all the IPv4 addresses will be returned by the inital `getaddrinfo()`, and the `ethtool` "fallback" will be used for getting the IPv6 addresses (if available).

Hence, I don't think this change makes a lot of difference in terms of "code flow". The only problem here is Python 3.9 crashing, and _I think_ it was fixed in stable branches of Python 3.10+ with https://github.com/python/cpython/pull/101252 -- I will need to check that out. Also, I'm surprised nobody actually had/reported issues when registering a RHEL 9 system without IPv6...

[1] https://man7.org/linux/man-pages/man3/getaddrinfo.3.html
